src/dscanner/reports.d

```diff
@@ -0,0 +1,160 @@
+//			Copyright Brian Schott (Hackerpilot) 2012.
+// Distributed under the Boost Software License, Version 1.0.
+//	  (See accompanying file LICENSE_1_0.txt or copy at
+//			http://www.boost.org/LICENSE_1_0.txt)
+
+module dscanner.reports;
+
+import std.json;
+import std.algorithm : map;
+import std.array : split, array;
+
+import dscanner.analysis.base : Message, MessageSet;
+
+class SonarQubeGenericIssueDataReporter
+{
+	enum Type
+	{
+		bug = "BUG",
+		vulnerability = "VULNERABILITY",
+		codeSmell = "CODE_SMELL"
+	}
+
+	enum Severity
+	{
+		blocker = "BLOCKER",
+		critical = "CRITICAL",
+		major = "MAJOR",
+		minor = "MINOR",
+		info = "INFO"
+	}
+
+	struct Issue
+	{
+		string engineId;
+		string ruleId;
+		Location primaryLocation;
+		string type;
+		string severity;
+		int effortMinutes;
+		Location[] secondaryLocations;
+	}
+
+	struct Location
+	{
+		string message;
+		string filePath;
+		TextRange textRange;
+	}
+
+	struct TextRange
+	{
+		// SonarQube Generic Issue only allows specifying start line only
+		// or the complete range, for which start and end has to differ
+		// D-Scanner does not provide the complete range info
+		long startLine;
+	}
+
+	private Issue[] _issues;
+
+	void addMessageSet(MessageSet messageSet)
+	{
+		_issues ~= toIssues(messageSet);
+	}
+	
+	void addMessage(Message message, bool isError = false)
+	{
+		_issues ~= toIssue(message, isError);
+	}
+
+	string getContent()
+	{
+		JSONValue result = [
+			"issues" : JSONValue(_issues.map!(e => toJson(e)).array)
+		];
+		return result.toString();
+	}
+
+	private JSONValue toJson(Issue issue)
```

Comment:
maybe all the private methods here should be made `static`?

Reply:
fixed